src/lib/mina_graphql/mina_graphql.ml

```diff
@@ -4255,6 +4255,90 @@ module Queries = struct
         Mina_lib.runtime_config mina
         |> Runtime_config.to_yojson |> Yojson.Safe.to_basic )
 
+  let get_epoch_ledgers ~mina breadcrumb =
+    let open Deferred.Result.Let_syntax in
+    let mina_config = Mina_lib.config mina in
+    let frontier_consensus_local_state = mina_config.consensus_local_state in
+    let consensus_state =
+      breadcrumb |> Transition_frontier.Breadcrumb.protocol_state
+      |> Mina_state.Protocol_state.consensus_state
+    in
+    let staking_epoch =
+      Consensus.Proof_of_stake.Data.Consensus_state.staking_epoch_data
+        consensus_state
+    in
+    let next_epoch =
+      Consensus.Proof_of_stake.Data.Consensus_state.next_epoch_data
+        consensus_state
+    in
+    let cast_ledger = function
+      | Consensus.Data.Local_state.Snapshot.Ledger_snapshot.Genesis_epoch_ledger
+          l ->
+          Ledger.Any_ledger.cast (module Ledger) l
+      | Consensus.Data.Local_state.Snapshot.Ledger_snapshot.Ledger_db l ->
+          Ledger.Any_ledger.cast (module Ledger.Db) l
+    in
+    let root_consensus_state =
+      let frontier =
+        Option.value_exn @@ Pipe_lib.Broadcast_pipe.Reader.peek
+        @@ Mina_lib.transition_frontier mina
+      in
+      let frontier_root = Transition_frontier.root frontier in
+      frontier_root |> Transition_frontier.Breadcrumb.protocol_state
+      |> Mina_state.Protocol_state.consensus_state
+    in
+    let%map staking_ledger, next_epoch_ledger =
+      match
+        (* We pretend that the block is finalized, so that we can query it in
+           advance, for redundancy.
+        *)
+        Consensus.Hooks.get_epoch_ledgers_for_finalized_frontier_block
+          ~root_consensus_state ~target_consensus_state:consensus_state
+          ~local_state:frontier_consensus_local_state
+      with
+      | `Both (staking_ledger, next_epoch_ledger) ->
+          return (cast_ledger staking_ledger, cast_ledger next_epoch_ledger)
+      | `Snarked_ledger (staking_ledger, num_parents) ->
+          (* The epoch transition was at a block between the current block and
+             the root. We find it by walking back by `num_parents` blocks.
+          *)
+          let%bind epoch_transition_state_hash =
+            let open Result.Let_syntax in
+            let rec ancestor breadcrumb i =
+              if i = 0 then return breadcrumb
```

Comment:
Is it intentional that we're trying to rollback to the first block of the epoch.

As I read the base case: for target block being the very first block of an epoch, it'll have `epoch_length = 1` and then `num_parents = 0` which would result in us retrieving the fully snarked ledger of the first block of this epoch.

Isn't it the last block of the previous epoch that we actually want?

Reply:
I misread the code, it seems. You're right: the code [here](https://github.com/MinaProtocol/mina/blob/30962d7c32142ce11479cc885b8fd1ff4c3f9929/src/lib/consensus/proof_of_stake.ml#L3217-L3229), which is called via [here](https://github.com/MinaProtocol/mina/blob/30962d7c32142ce11479cc885b8fd1ff4c3f9929/src/lib/transition_frontier/full_frontier/full_frontier.ml#L404) using the snarked ledger from [here](https://github.com/MinaProtocol/mina/blob/30962d7c32142ce11479cc885b8fd1ff4c3f9929/src/lib/transition_frontier/transition_frontier.ml#L450), computed before we've transitioned. I'll update it.